panic: runtime error: index out of range at Split()

Thank you for the great library!

I have a crash when use the Split() function, below is the details:

**Version**: main
**Steps to reproduce**

```
func main() {
    data := make([]byte, 512)
    var ecctest reedsolomon.Encoder
    ecctest, _ = reedsolomon.New(1, 0)
    _, _ = ecctest.Split(data)
}
```

**Log**
```

panic: runtime error: index out of range [512] with length 512

goroutine 1 [running]:
github.com/klauspost/reedsolomon.(*reedSolomon).Split(0x1?, {0xc0000a6000, 0x0?, 0x0?})
```

I have read the code, and find the issues may from this code:

https://github.com/klauspost/reedsolomon/blob/b413df32aec2953ebc9e85167dc4b4969be4a3a6/reedsolomon.go#L1564

When `len(data) == (r.totalShards * perShard)` it will go to line 1577, and try to set `data[512]` as `0` which will cause index out of range. It seems no need to do if `len(data) == (r.totalShards * perShard)` ?

Thanks!

--- a/leopard.go
+++ b/leopard.go
@@ -272,6 +272,9 @@ func (r *leopardFF16) Split(data []byte) ([][]byte, error) {
 	if len(data) == 0 {
 		return nil, ErrShortData
 	}
+	if r.totalShards == 1 && len(data)&63 == 0 {
+		return [][]byte{data}, nil
+	}
 	dataLen := len(data)
 	// Calculate number of bytes per data shard.
 	perShard := (len(data) + r.dataShards - 1) / r.dataShards
@@ -295,8 +298,9 @@ func (r *leopardFF16) Split(data []byte) ([][]byte, error) {
 			copyFrom = copyFrom[copy(padding[i], copyFrom):]
 		}
 	} else {
-		for i := dataLen; i < dataLen+r.dataShards; i++ {
-			data[i] = 0
+		zero := data[dataLen : r.totalShards*perShard]
+		for i := range zero {
+			zero[i] = 0
 		}
 	}
 

--- a/leopard8.go
+++ b/leopard8.go
@@ -312,6 +312,10 @@ func (r *leopardFF8) Split(data []byte) ([][]byte, error) {
 	if len(data) == 0 {
 		return nil, ErrShortData
 	}
+	if r.totalShards == 1 && len(data)&63 == 0 {
+		return [][]byte{data}, nil
+	}
+
 	dataLen := len(data)
 	// Calculate number of bytes per data shard.
 	perShard := (len(data) + r.dataShards - 1) / r.dataShards
@@ -335,8 +339,9 @@ func (r *leopardFF8) Split(data []byte) ([][]byte, error) {
 			copyFrom = copyFrom[copy(padding[i], copyFrom):]
 		}
 	} else {
-		for i := dataLen; i < dataLen+r.dataShards; i++ {
-			data[i] = 0
+		zero := data[dataLen : r.totalShards*perShard]
+		for i := range zero {
+			zero[i] = 0
 		}
 	}
 

--- a/reedsolomon.go
+++ b/reedsolomon.go
@@ -1551,6 +1551,10 @@ func (r *reedSolomon) Split(data []byte) ([][]byte, error) {
 	if len(data) == 0 {
 		return nil, ErrShortData
 	}
+	if r.totalShards == 1 {
+		return [][]byte{data}, nil
+	}
+
 	dataLen := len(data)
 	// Calculate number of bytes per data shard.
 	perShard := (len(data) + r.dataShards - 1) / r.dataShards
@@ -1574,8 +1578,9 @@ func (r *reedSolomon) Split(data []byte) ([][]byte, error) {
 		}
 		data = data[0 : perShard*fullShards]
 	} else {
-		for i := dataLen; i < dataLen+r.dataShards; i++ {
-			data[i] = 0
+		zero := data[dataLen : r.totalShards*perShard]
+		for i := range zero {
+			zero[i] = 0
 		}
 	}
 

--- a/reedsolomon_test.go
+++ b/reedsolomon_test.go
@@ -1950,6 +1950,20 @@ func TestNew(t *testing.T) {
 	}
 }
 
+func TestSplitZero(t *testing.T) {
+	data := make([]byte, 512)
+	for _, opts := range testOpts() {
+		ecctest, err := New(1, 0, opts...)
+		if err != nil {
+			t.Fatal(err)
+		}
+		_, err = ecctest.Split(data)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
 // Benchmark 10 data shards and 4 parity shards and 160MB data.
 func BenchmarkSplit10x4x160M(b *testing.B) {
 	benchmarkSplit(b, 10, 4, 160*1024*1024)

